### PR TITLE
types: revoke #19157, compare nil by type assertion in datum value setting

### PR DESCRIPTION
**Pull request by Hexilee**
Signed-off-by: hexilee <i@hexilee.me>

### What problem does this PR solve?

Problem Summary:

The #19157 cannot resolve the problem in its description, 
because an `interface{}` referring to an object containing a nil ptr is not equal to nil.

For example, following code will print 'false':

```go
var a interface{} = []byte(nil)
fmt.Println(a == nil)
```

What's Changed:

`types.Datum.SetValue` and `types.Datum.SetValueWithDefaultCollation`

### Release note 

- No release note

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -430,11 +430,9 @@ func (d *Datum) GetValue() interface{} {
 
 // SetValueWithDefaultCollation sets any kind of value.
 func (d *Datum) SetValueWithDefaultCollation(val interface{}) {
-	if val == nil {
-		d.SetNull()
-		return
-	}
```
Comment (qw4990) [inline types/datum.go:436]: Should we also keep these lines?
If the `val` is nil, would type assertion on it cause panic?

Comment (Hexilee) [reply types/datum.go:436]: It won't cause panic, as [example](https://play.golang.org/p/M44mnWePb6C).

**Comment by Hexilee** [comment]
> Please add some tests for this fix.

Are tests really necessary? They have existed for a long time, this PR just roll them back to an earlier version before #19175. This is the [earlier version](https://github.com/pingcap/tidb/blob/030eab2bca2980245af8386d622b8852273bee7e/types/datum.go#L432).

**Pull request merged**